/*
 * Copyright 2018 The Feast Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package feast.core;

import "google/protobuf/timestamp.proto";

option java_package = "feast.core";
option java_outer_classname = "DatasetServiceProto";
option go_package = "github.com/gojek/feast/protos/generated/go/feast/core";

service DatasetService {
  // Create training dataset for a feature set
  rpc CreateDataset(DatasetServiceTypes.CreateDatasetRequest) returns (DatasetServiceTypes.CreateDatasetResponse){};
}

message DatasetServiceTypes {
  message CreateDatasetRequest {
    // set of features for which its training data should be created
    FeatureSet featureSet = 1;
    // start date of the training data (inclusive)
    google.protobuf.Timestamp startDate = 2;
    // end date of the training data (inclusive)
    google.protobuf.Timestamp endDate = 3;
    // (optional) number of row that should be generated
    // (default: none)
    int64 limit = 4;
    // (optional) prefix for dataset name
    string namePrefix = 5;
  }

  message CreateDatasetResponse {
    // information of the created training dataset
    DatasetInfo datasetInfo = 1;
  }
}

// Represent a collection of feature having same entity name
message FeatureSet {
  // entity related to this feature set
  string entityName = 1;
  // list of feature id in this feature set
  repeated string featureIds = 2;
}

// Representation of training dataset information
message DatasetInfo {
  // name of dataset
  string name = 1;
  // URL to table location of the training dataset
  string tableUrl = 2;
}